unixfs/src/walk.rs

```diff
@@ -639,216 +642,73 @@ impl InnerEntry {
 }
 
-/// Errors which can occur while walking a tree.
 #[derive(Debug)]
-pub enum Error {
-    /// An unsupported type of UnixFS node was encountered. There should be a way to skip these. Of the
-    /// defined types only `Metadata` is unsupported, all undefined types as of 2020-06 are also
-    /// unsupported.
-    UnsupportedType(UnexpectedNodeType),
-
-    /// This error is returned when a file e.g. links to a non-Raw or non-File subtree.
-    UnexpectedType(UnexpectedNodeType),
-
-    /// dag-pb node parsing failed, perhaps the block is not a dag-pb node?
-    DagPbParsingFailed(quick_protobuf::Error),
-
-    /// Failed to parse the unixfs node inside the dag-pb node.
-    UnixFsParsingFailed(quick_protobuf::Error),
-
-    /// dag-pb node contained no data.
-    EmptyDagPbNode,
-
-    /// dag-pb link could not be converted to a Cid
-    InvalidCid(InvalidCidInLink),
-
-    /// A File has an invalid structure
-    File(FileError),
-
-    /// A Directory has an unsupported structure
-    UnsupportedDirectory(UnexpectedDirectoryProperties),
-
-    /// HAMTSharded directory has unsupported properties
-    UnsupportedHAMTShard(ShardError),
-}
-
-impl From<ParsingFailed<'_>> for Error {
-    fn from(e: ParsingFailed<'_>) -> Self {
-        use ParsingFailed::*;
-        match e {
-            InvalidDagPb(e) => Error::DagPbParsingFailed(e),
-            InvalidUnixFs(e, _) => Error::UnixFsParsingFailed(e),
-            NoData(_) => Error::EmptyDagPbNode,
-        }
-    }
-}
-
-impl From<InvalidCidInLink> for Error {
-    fn from(e: InvalidCidInLink) -> Self {
-        Error::InvalidCid(e)
-    }
-}
-
-impl From<FileReadFailed> for Error {
-    fn from(e: FileReadFailed) -> Self {
-        use FileReadFailed::*;
-        match e {
-            File(e) => Error::File(e),
-            UnexpectedType(ut) => Error::UnexpectedType(ut),
-            Read(_) => unreachable!("FileVisit does not parse any blocks"),
-            InvalidCid(l) => Error::InvalidCid(l),
-        }
-    }
-}
-
-impl From<UnexpectedDirectoryProperties> for Error {
-    fn from(e: UnexpectedDirectoryProperties) -> Self {
-        Error::UnsupportedDirectory(e)
-    }
+// FIXME: could simplify roots to optinal cid variants?
+enum InnerKind {
+    /// This is necessarily at the root of the walk
+    RootDirectory,
+    /// This is necessarily at the root of the walk
+    BucketAtRoot,
+    /// This is the metadata containing bucket, for which we have a name
+    RootBucket,
+    /// This is a sibling to a previous named metadata containing bucket
+    Bucket,
+    /// Directory on any level except root
+    Directory,
+    /// File optionally on the root level
+    File(Option<FileVisit>, u64),
+    /// Symlink optionally on the root level
+    Symlink,
 }
 
-impl From<ShardError> for Error {
-    fn from(e: ShardError) -> Self {
-        Error::UnsupportedHAMTShard(e)
-    }
+/// Converts a link of specifically a Directory (and not a link of a HAMTShard).
+fn convert_link(
```

Comment:
Moving these link converters back where they used to be would be a big win for the lines changed.

Reply:
It is usual to keep private elements after public ones and helps with organization and code visibility. For example, many projects like Rustc itself and rust-analyzer also do this.

```
1. fn foo();
2. pub struct A;
3. pub enum B;
4. enum C;
5. pub fn bar();
```
->
```
1. pub struct A;
2. pub enum B;
3. pub fn bar();
4. fn foo();
5. enum C;
```

Reply:
That's all fine; however, this kind of change (formatting, ordering) should be separated from functional changes, i.e. contained in its own dedicated commit, so that the functional changes can be reviewed separately from visual ones.

Reply:
It is possible to see this ordering solely in the first commit. The other two were order restrained by the maximum of my capacity

Reply:
> It is possible to see this ordering solely in the first commit.
That's true, though due to the fact that it is the first commit and the later ones build on top of it, it is not possible to reverse it to view the whole changeset without it.

In addition, in this case the size of the changes introduced by the reordering is quite big, warranting its very own PR. We have updated our [contributing guidelines](https://github.com/rs-ipfs/rust-ipfs/blob/master/CONTRIBUTING.md) to make it more clear: the patch size should be kept minimal.